README.md
# Helix Index Big Tree

> Like [`helix-index-tree`](https://github.com/adobe/helix-index-tree), but for big trees

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-index-big-tree.svg)](https://codecov.io/gh/adobe/helix-index-big-tree)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-index-big-tree.svg)](https://circleci.com/gh/adobe/helix-index-big-tree)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-index-big-tree.svg)](https://github.com/adobe/helix-index-big-tree/blob/master/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-index-big-tree.svg)](https://github.com/adobe/helix-index-big-tree/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-index-big-tree.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-index-big-tree)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

## Usage

```bash
wsk action invoke helix-services/index-big-tree@v1
```

This service is meant to be invoked from [helix-index-tree](https://github.com/adobe/helix-index-tree), so it won't be made available as a web action.

For more, see the [API documentation](docs/API.md).

## Development

### Deploying Helix Service

Deploying Helix Service requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix-index` namespace.

All commits to master that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-index/index-big-tree@ci<num>` and tagged with the CI build number.


CHANGELOG.md
## [1.0.32](https://github.com/adobe/helix-index-big-tree/compare/v1.0.31...v1.0.32) (2020-07-02)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v8.1.0 ([#118](https://github.com/adobe/helix-index-big-tree/issues/118)) ([f649c6d](https://github.com/adobe/helix-index-big-tree/commit/f649c6dee2a19141ecd742e00dd5dc060e09c1df))

## [1.0.31](https://github.com/adobe/helix-index-big-tree/compare/v1.0.30...v1.0.31) (2020-07-01)


### Bug Fixes

* **deps:** update dependency epsagon to v1.73.4 ([0a31b69](https://github.com/adobe/helix-index-big-tree/commit/0a31b6993bb99645b0b033cb5b0660c16f88921a))

## [1.0.30](https://github.com/adobe/helix-index-big-tree/compare/v1.0.29...v1.0.30) (2020-07-01)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v8 ([8b08dd1](https://github.com/adobe/helix-index-big-tree/commit/8b08dd16cabfeffe58f8aa2d25d37d989b792128))

## [1.0.29](https://github.com/adobe/helix-index-big-tree/compare/v1.0.28...v1.0.29) (2020-07-01)


### Bug Fixes

* **deps:** update dependency epsagon to v1.73.3 ([#114](https://github.com/adobe/helix-index-big-tree/issues/114)) ([8302232](https://github.com/adobe/helix-index-big-tree/commit/8302232200c67b92a18fbc637e3cdd81eeb2f210))

## [1.0.28](https://github.com/adobe/helix-index-big-tree/compare/v1.0.27...v1.0.28) (2020-06-30)


### Bug Fixes

* **deps:** update external ([b869906](https://github.com/adobe/helix-index-big-tree/commit/b8699061beda6fc3658122234d3f5475567fa1f8))
* **deps:** update external ([#112](https://github.com/adobe/helix-index-big-tree/issues/112)) ([8b24eb7](https://github.com/adobe/helix-index-big-tree/commit/8b24eb7fd05773052a39b5c30acf5fe63b490fc7))

## [1.0.27](https://github.com/adobe/helix-index-big-tree/compare/v1.0.26...v1.0.27) (2020-06-09)


### Bug Fixes

* **deps:** update external ([#107](https://github.com/adobe/helix-index-big-tree/issues/107)) ([c074f9f](https://github.com/adobe/helix-index-big-tree/commit/c074f9fc9e4986531858a65912e16c1eb7ad4c1b))

## [1.0.26](https://github.com/adobe/helix-index-big-tree/compare/v1.0.25...v1.0.26) (2020-06-02)


### Bug Fixes

* **deps:** update external ([#103](https://github.com/adobe/helix-index-big-tree/issues/103)) ([5e265e5](https://github.com/adobe/helix-index-big-tree/commit/5e265e5fe6f32b4d1ef12629e77b3749747708df))

## [1.0.25](https://github.com/adobe/helix-index-big-tree/compare/v1.0.24...v1.0.25) (2020-04-16)


### Bug Fixes

* **deps:** update [@adobe](https://github.com/adobe) ([#86](https://github.com/adobe/helix-index-big-tree/issues/86)) ([4f77ec4](https://github.com/adobe/helix-index-big-tree/commit/4f77ec4101c4cf9d11f729024253762136a08ef3))

## [1.0.24](https://github.com/adobe/helix-index-big-tree/compare/v1.0.23...v1.0.24) (2020-04-14)


### Bug Fixes

* **deploy:** use helix-services package ([#85](https://github.com/adobe/helix-index-big-tree/issues/85)) ([8d80872](https://github.com/adobe/helix-index-big-tree/commit/8d808727601f46d7bf1405387c5a740147fa918e))

## [1.0.23](https://github.com/adobe/helix-index-big-tree/compare/v1.0.22...v1.0.23) (2020-04-07)


### Bug Fixes

* **deps:** update external ([#81](https://github.com/adobe/helix-index-big-tree/issues/81)) ([666421f](https://github.com/adobe/helix-index-big-tree/commit/666421ffd87f4ba41e868cc7ed20f6d4a1b65f21))

## [1.0.22](https://github.com/adobe/helix-index-big-tree/compare/v1.0.21...v1.0.22) (2020-03-17)


### Bug Fixes

* **deps:** update external ([#67](https://github.com/adobe/helix-index-big-tree/issues/67)) ([f1cf302](https://github.com/adobe/helix-index-big-tree/commit/f1cf302f601c372901b730e1e37006e73d5211d1))

## [1.0.21](https://github.com/adobe/helix-index-big-tree/compare/v1.0.20...v1.0.21) (2020-03-10)


### Bug Fixes

* **deps:** update dependency tar-stream to v2.1.1 ([#65](https://github.com/adobe/helix-index-big-tree/issues/65)) ([4020703](https://github.com/adobe/helix-index-big-tree/commit/4020703fb017d4d5b20c1a4bc45f576374163c99))

## [1.0.20](https://github.com/adobe/helix-index-big-tree/compare/v1.0.19...v1.0.20) (2020-03-02)


### Bug Fixes

* **deps:** update external ([b93375d](https://github.com/adobe/helix-index-big-tree/commit/b93375dfac2aa763be75f84955a89dff607c73a7))

## [1.0.19](https://github.com/adobe/helix-index-big-tree/compare/v1.0.18...v1.0.19) (2020-02-25)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v4.5.1 ([0f4070b](https://github.com/adobe/helix-index-big-tree/commit/0f4070bff1821788b03aa8f25a4f17dc42060199))
* **deps:** update dependency @adobe/helix-status to v7.1.2 ([3dac09e](https://github.com/adobe/helix-index-big-tree/commit/3dac09e6fd639a725be25e7210d74a3c3b622b09))

## [1.0.18](https://github.com/adobe/helix-index-big-tree/compare/v1.0.17...v1.0.18) (2020-02-18)


### Bug Fixes

* **deps:** update external ([#56](https://github.com/adobe/helix-index-big-tree/issues/56)) ([e19160b](https://github.com/adobe/helix-index-big-tree/commit/e19160be1955fa15795ed42f0d62b7eb12185485))

## [1.0.17](https://github.com/adobe/helix-index-big-tree/compare/v1.0.16...v1.0.17) (2020-01-24)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v7.1.1 ([5e9f75a](https://github.com/adobe/helix-index-big-tree/commit/5e9f75a475e328299a353a877157ec83c095e10b))

## [1.0.16](https://github.com/adobe/helix-index-big-tree/compare/v1.0.15...v1.0.16) (2020-01-22)


### Bug Fixes

* **deps:** update [@adobe](https://github.com/adobe) ([5d4dcc0](https://github.com/adobe/helix-index-big-tree/commit/5d4dcc03ae923650dd46a4b968da4eefb1a34b2c))

## [1.0.15](https://github.com/adobe/helix-index-big-tree/compare/v1.0.14...v1.0.15) (2020-01-14)


### Bug Fixes

* **deps:** update external ([#37](https://github.com/adobe/helix-index-big-tree/issues/37)) ([8f15cfd](https://github.com/adobe/helix-index-big-tree/commit/8f15cfd92586b6dde3754db67f21a0fd540b2e71))

## [1.0.14](https://github.com/adobe/helix-index-big-tree/compare/v1.0.13...v1.0.14) (2020-01-09)


### Bug Fixes


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

'use strict';

const assert = require('assert');
const proxyquire = require('proxyquire');
const sinon = require('sinon');
// const NodeHttpAdapter = require('@pollyjs/adapter-node-http');
// const FSPersister = require('@pollyjs/persister-fs');
// const setupPolly = require('@pollyjs/core').setupMocha;

describe('Index Tests', () => {
  /*
  // polly screws up the recording of tar.gz downloads. enable at your own risk
  setupPolly({
    logging: false,
    recordFailedRequests: true,
    recordIfMissing: true,
    adapters: [NodeHttpAdapter],
    persister: FSPersister,
    persisterOptions: {
      fs: {
        recordingsDir: 'test/fixtures',
      },
    },
    matchRequestsBy: {
      headers: {
        exclude: ['authorization', 'user-agent'],
      },
    },
  });
  */

  let index;
  let invoke;
  beforeEach(() => {
    invoke = sinon.fake();
    index = proxyquire('../src/index.js', {
      openwhisk: () => ({
        actions: {
          invoke,
        },
      }),
    }).main;
  });

  it('index function bails if necessary arguments are missing', async () => {
    const res = await index();
    assert.equal(res.statusCode, 400);
    assert.equal(res.body, 'Required arguments missing');
  });

  it('index function makes HTTP requests', async () => {
    const batchsize = 1;
    const result = await index({
      owner: 'trieloff',
      repo: 'helix-demo',
      ref: 'e266e69024853cc6b25fdcfb963d2d0014162f1c',
      branch: 'master',
      batchsize,
    });
    assert.equal(typeof result, 'object');
    assert.deepEqual(result, {
      statusCode: 201,
      delegated: 'update-index',
      failures: [],
      jobs: 1,
    });

    sinon.assert.callCount(invoke, 1);
  }).timeout(2000);

  it('index function makes HTTP requests with custom batchsize', async () => {
    const batchsize = 10;
    const result = await index({
      owner: 'trieloff',
      repo: 'helix-demo',
      ref: 'e266e69024853cc6b25fdcfb963d2d0014162f1c',
      branch: 'master',
      batchsize,
    });
    assert.equal(typeof result, 'object');
    assert.deepEqual(result, {
      statusCode: 201,
      delegated: 'update-index',
      failures: [],
      jobs: 1,
    });

    sinon.assert.callCount(invoke, 1);
  }).timeout(2000);

  it('index filters by pattern', async () => {
    const result = await index({
      owner: 'trieloff',
      repo: 'helix-demo',
      ref: 'ca8959afbb2668c761e47a4563f054da2444ab30',
      branch: 'master',
      pattern: '**/*.{md,html}',
      batchsize: 10,
    });
    assert.equal(typeof result, 'object');
    assert.deepEqual(result, {
      statusCode: 201,
      failures: [],
      delegated: 'update-index',
      jobs: 7,
    });
    sinon.assert.callCount(invoke, 1);
  }).timeout(50000);

  it.skip('index super large repo', async () => {
    const result = await index({
      owner: 'MicrosoftDocs',
      repo: 'azure-docs',
      ref: 'a28fc7ad76fcbf92cbdcba7f2908ec1226e494ad',
      branch: 'master',
      pattern: '**/*.md',
      batchsize: 1000,
    });
    assert.equal(typeof result, 'object');
    assert.deepEqual(result, {
      delegated: 'update-index',
      jobs: 1451,
    });
    sinon.assert.callCount(invoke, result.body.jobs);
  }).timeout(5000000);
});


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

const {
  info, error, rootLogger, CoralogixLogger,
} = require('@adobe/helix-log');
const { wrap } = require('@adobe/helix-status');
const phin = require('phin');
const minimatch = require('minimatch');
const openwhisk = require('openwhisk');
const tar = require('tar-stream');
const { createGunzip } = require('gunzip-stream');
const strip = require('strip-dirs');

/**
 * This is the main function
 * @param {string} name name of the person to greet
 * @returns {object} a greeting
 */
async function main({
  owner, repo, ref, branch, pattern = '**/*.{md,jpg}', token, CORALOGIX_TOKEN, batchsize = 1,
} = {}) {
  rootLogger.loggers.set('coralogix', new CoralogixLogger(
    CORALOGIX_TOKEN, 'test', '/trieloff/helix-index-big-tree',
  ));

  if (!(owner && repo && ref && branch)) {
    error('Required arguments missing');
    return {
      statusCode: 400,
      body: 'Required arguments missing',
    };
  }

  const ow = openwhisk();

  try {
    const prom = new Promise((resolve) => {
      const list = tar.extract();
      const jobs = [];
      const failures = [];
      let batch = [];

      let lastpt = new Date().getTime();
      let count = 0;
      list.on('entry', async (header, stream, next) => {
        const path = strip(header.name, 1);

        if (minimatch(path, pattern)) {
          count += 1;
          if (new Date().getTime() - lastpt > 5000) {
            info('batching #', count, path);
            lastpt = new Date().getTime();
          }
          batch.push(path);
          if (batch.length >= Number.parseInt(batchsize, 10)) {
            const paths = batch.slice();
            // start a new batch
            batch = [];

            info(`invoking #${count} with ${paths.length} items`);
            try {
              ow.actions.invoke({
                name: 'helix-index/index-file@1.2.1',
                blocking: false,
                result: false,
                params: {
                  owner, repo, ref, path, paths, branch, sha: 'initial', token,
                },
              });
              jobs.push(...paths);
            } catch (e) {
              error(e);
              failures.push(...paths);
            }
          }
        }

        stream.on('end', () => {
          next();
        });

        stream.resume();
      });

      list.on('finish', async () => {
        info('tar finished');

        const paths = batch.slice();
        // start a new batch
        batch = [];
        info('batch complete', paths.length);
        if (paths.length > 0) {
          info(`invoking final batch with ${paths.length} items`);
          try {
            await ow.actions.invoke({
              name: 'helix-index/index-file@1.2.1',
              blocking: false,
              result: false,
              params: {
                owner, repo, ref, paths, branch, sha: 'initial', token,
              },
            });
            jobs.push(...paths);
          } catch (e) {
            error(e);
            failures.push(...paths);
          }
        }

        resolve({
          statusCode: 201,
          delegated: 'update-index',
          jobs: jobs.length,
          failures,
        });
      });

      phin({
        url: `https://github.com/${owner}/${repo}/tarball/${ref}`,
        stream: true,
        followRedirects: true,
      }).then((res) => {
        res
          .pipe(createGunzip())
          .pipe(list);
      });
    });

    info('download started');

    const res = await prom;
    info('completed');
    return res;
  } catch (e) {
    error(e);
    return {
      statusCode: 500,
      body: e.message,
    };
  }
}

module.exports = { main: wrap(main) };
